Implemented connecting to the rooms.

Subv opened the pull request:
Nicknames and MAC addresses are not yet negotiated.
Only wifi beacon frames are retransmitted for the moment.

Moved the header files to a tunnler/ subdirectory.

There are still some concurrency issues that i'd like help with, namely in RoomMember.

Subv committed:
Implemented connecting to the rooms.

Nicknames and MAC addresses are not yet negotiated.
Only wifi beacon frames are retransmitted for the moment.

Moved the header files to a tunnler/ subdirectory

Subv commented on include/tunnler/room_member.h:
```diff
@@ -108,31 +112,41 @@ class RoomMember final {
-    State state;    // Current state of the RoomMember
-    MemberInformation member_information;
-    RoomInformation room_information;
+    State state; ///< Current state of the RoomMember.
```
Maybe wrap this in std::atomic?

JayFoxRox commented on include/tunnler/room.h:
```diff
@@ -5,47 +5,56 @@
 
-    /** 
+    /**
      * Gets the current status of the room.
```
nit: If you rename it to `GetState` also say "state" here

Subv commented on src/room_member.cpp:
```diff
@@ -0,0 +1,196 @@
+
+void RoomMember::ReceiveLoop() {
+    // Receive packets while the connection is open
+    while (IsConnected()) {
```
I think this is a race condition with Leave()

JayFoxRox commented on include/tunnler/room_member.h:
```diff
@@ -86,13 +90,13 @@ class RoomMember final {
      * @param packet The WiFi packet to send.
      */
-    void SendWifiPacket(const WifiPacket& packet);
+    void SendWifiPacket(WifiPacket& packet);
```
Why isn't this const anymore?

Subv replied:
Because there's a very very awful
````
stream.Write(reinterpret_cast<char*>(wifi_packet.data.data()), wifi_packet.data.size());
````

which would have to become either
````
stream.Write(const_cast<char*>(reinterpret_cast<const char*>(wifi_packet.data.data())), wifi_packet.data.size());
````
or
````
stream.Write((char*)wifi_packet.data.data(), wifi_packet.data.size());
````

Suggestions?

JayFoxRox replied:
I'd do the casting back and forth or even create a copy of the packet, just to not give in to RakNet shenannigans.

Before sending that E-Mail I might check if there is a PR / reason for this and tell them to fix it too :P

JayFoxRox replied:
To resolve this before the merge: please make it accept a const ref. I feel like this is part of a stable API so it would suck to change it later

JayFoxRox commented on include/tunnler/room_member.h:
```diff
@@ -108,31 +112,41 @@ class RoomMember final {
+    static const size_t MaxBeaconQueueSize = 25;
+
+    std::deque<ChatEntry> chat_queue;    ///< List of all chat messages recieved since last PopChatEntries was called
+    std::deque<WifiPacket> data_queue;   ///< List of all recieve 802.11 frames with type Data
```
spelling: "received" - Maybe also put the word "Data" in backticks / quotes?
Same with Beacon line below

B3n30 commented on include/tunnler/room.h:
```diff
@@ -5,47 +5,56 @@
 
-    /** 
+    /**
      * Creates the socket for this room. Returns true on success. Will bind to default address if server is empty string.
```
Doesn't return anything

JayFoxRox commented on src/room.cpp:
```diff
@@ -0,0 +1,59 @@
+#include "tunnler/room_message_types.h"
+
+/// Maximum number of concurrent connections allowed to this room.
+static const uint32_t MaxConcurrentConnections = 10;
```
In the long run this should be moved to the server creation / memberlimit. Temporarily it's fine

JayFoxRox commented on src/room.cpp:
```diff
@@ -0,0 +1,59 @@
+        std::lock_guard<std::mutex> lock(server_mutex);
+
+        if (server)
+            RakNet::RakPeerInterface::DestroyInstance(server);
```
explicit braces / scope plox.

JayFoxRox commented on src/room.cpp:
```diff
@@ -0,0 +1,59 @@
+            RakNet::RakPeerInterface::DestroyInstance(server);
+        server = nullptr;
+    }
+
```
We should assert that we closed successfully. I'm not sure if raknet destroying the instance returns anything / could fail. But it would be quite horrendous if a host closes the server, returns a "Closed" state but might still have some socket open.

If raknet is fail-safe in this regard this is fine.

Subv replied:
I have no idea if that's the case, what would we even check for here?

JayFoxRox commented on src/room.cpp:
```diff
@@ -0,0 +1,59 @@
+
+        RakNet::Packet* packet = server->Receive();
+        if (!packet)
+            continue;
```
explicit scope once more

JayFoxRox commented on src/room.cpp:
```diff
@@ -0,0 +1,59 @@
+            server->Send(reinterpret_cast<char*>(packet->data), packet->length,
+                         HIGH_PRIORITY, RELIABLE_ORDERED, 0, packet->systemAddress, true);
+            break;
+        }
```
This should have a fallback ASAP so we see any raknet packet we might have to worry about in the future

Subv replied:
what would the fallback do? we'll be getting *tons* of raknet messages that we don't care about (incoming_connection notification, etc)

JayFoxRox commented on src/room.cpp:
```diff
@@ -0,0 +1,59 @@
+        case ID_ROOM_WIFI_PACKET:
+            // Received a wifi packet, broadcast it to everyone else except the sender.
+            server->Send(reinterpret_cast<char*>(packet->data), packet->length,
+                         HIGH_PRIORITY, RELIABLE_ORDERED, 0, packet->systemAddress, true);
```
nit: didn't we decide to make these `RELIABLE` but unordered?

Subv replied:
true, leftover from testing

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,196 @@
+
+RoomMember::~RoomMember() {
+    if (receive_thread)
+        receive_thread->join();
```
more scope fun

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,196 @@
+
+        // If we have more than the max number of buffered beacons, discard the oldest one
+        if (beacon_queue.size() > MaxBeaconQueueSize)
+            beacon_queue.pop_front();
```
*cough* - this time it actually even confused me because the indentation is so unexpected

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,196 @@
+        if (beacon_queue.size() > MaxBeaconQueueSize)
+            beacon_queue.pop_front();
+    } else {
+        // TODO(Subv): Add to ringbuffer for data / non-beacons
```
Sounds like we should just create our own class / struct for this?

Subv replied:
that's imo too much work for little payoff, we can just pop_front() the data queue like we do for the beacon queue

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,196 @@
+    stream.Write(static_cast<uint32_t>(wifi_packet.data.size()));
+    stream.Write(reinterpret_cast<char*>(wifi_packet.data.data()), wifi_packet.data.size());
+
+    peer->Send(&stream, HIGH_PRIORITY, RELIABLE_ORDERED, 0, server_address, false);
```
same thing again: not sure if RELIABLE only?

Subv committed:
Make wifi packets reliable but unordered.

We'll need to evaluate if this behavior is acceptable for the users of the interface (UDS for now)

Adhere to the correct codestyle.

Subv commented:
Addressed some @JayFoxRox comments.

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,196 @@
+            // Update the server address with the address of the sender of this packet.
+            server_address = packet->systemAddress;
+            // TODO(Subv): Send a room join request to the server before marking us as joined.
+            state = State::Joined;
```
Aren't we still Joining?
We are now halfway through the handshake.
I'd expect us to be completly joined once we have received confirmation for our nickname + mac address

Subv replied:
See the TODO. This was basically testing code.

JayFoxRox replied:
The TODO comment only says to send the packet. What I'm saying is: wait for a response.
So the state should not be modified in this function at all.

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,196 @@
+        case ID_CONNECTION_LOST:
+            // Couldn't deliver a reliable packet, the other system was abnormally terminated
+            state = State::LostConnection;
+            break;
```
Did you check which events are automaticly closing the connection and which don't?
If the server is full I'd expect raknet to internally have closed the connection. if it doesn't it might not do so here either? (maybe it will just try to reconnect when continuing sending data?)

Subv committed:
Shutdown the network connection when closing down or receiving an error message.

Subv committed:
Use an atomic variable to store the state of the RoomMeber and Room, as it is accessed by multiple threads concurrently.

fixup! Shutdown the network connection when closing down or receiving an error message.

fixup! Implemented connecting to the rooms.

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -23,9 +23,11 @@ RoomMember::RoomMember() {
 }
 
 RoomMember::~RoomMember() {
+    ASSERT_MSG(!IsConnected(), "RoomMeber is being destroyed while connected");
```
spelling

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,203 @@
+}
+
+RoomMember::~RoomMember() {
+    ASSERT_MSG(!IsConnected(), "RoomMeber is being destroyed while connected");
```
Spelling.. still [github is being stupid]

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,203 @@
+static const MacAddress NoPreferredMac = { 0, 0, 0, 0, 0, 0 };
+
+RoomMember::RoomMember() {
+    peer = RakNet::RakPeerInterface::GetInstance();
```
I believe offline raknet peers do some things such as responding to pings.
If that is the case, we should consider only opening the peers once actual connection stuff is about to happen maybe?

Subv committed:
Process all available packets on each network loop iteration.

Subv commented on src/room_member.cpp:
```diff
@@ -0,0 +1,200 @@
+void RoomMember::Leave() {
+    ASSERT_MSG(receive_thread != nullptr, "Must be in a room to leave it.");
+
+    peer->Shutdown(300);
```
This Shutdown() call will cause the peer->Receive() loop to stop immediately, is that desired? Should we care about queued packets?

Subv replied:
A possible fix would be to have a lock here and inside the while (IsConnected()) loop. That ensures that we have processed all queued packets before shutting down, similar to how it's done in the Room class.

Subv committed:
Hold a lock while receiving packets in the client to avoid race conditions between packet processing and room leaving.

fixup! Hold a lock while receiving packets in the client to avoid race conditions between packet processing and room leaving.

Subv committed:
Implement the initial join handshake.

The assigned MAC for all members is currently 0.
The ID_ROOM_JOIN_SUCCESS message does not currently send any information, and thus will cause the clients to fail an assert when receiving the join result.

fixup cmake to build with C++14, set NoPreferredMac = { 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF }

B3n30 commented on src/room.cpp:
```diff
@@ -0,0 +1,188 @@
+        std::lock_guard<std::mutex> lock(server_mutex);
+
+        if (server) {
+            server->Shutdown(300);
```
Shouldn't we send a ID_DISCONNECTION_NOTIFICATION to the clients?

JayFoxRox replied:
isn't that part of Shutdown? *assumes*

B3n30 replied:
You are right. Just checked the doc

B3n30 commented on src/room.cpp:
```diff
@@ -0,0 +1,188 @@
+        server = nullptr;
+    }
+
+    room_thread->join();
```
Shouldn't this be before the server is closed. I know for now it won't make a difference, but maybe for later implementations. It's just logically to end the ServerLoop before destroying the server.

JayFoxRox replied:
the room_thread is a `unique_ptr` / doesn't the thread get destroyed automaticly?
This is only do to things gracefully - we already assert in the other case so API users should know that shit is about to go down (we should just document the fact that rooms must be `destroy()`d before destroying the object).
However, maybe we *must* forcefully [not the same as `leave()`!] kill the thread to avoid race conditions where the room_thread will use an already closed / destroyed RakPeer - but we might be doing that already, I didn't check

B3n30 commented on src/room.cpp:
```diff
@@ -0,0 +1,188 @@
+
+void Room::ServerLoop() {
+    while (state != State::Closed) {
+        std::lock_guard<std::mutex> lock(server_mutex);
```
In the Raknet examples a Sleep(30 / u_sleep(30*1000) is used to "keep Raknet responsive". I don't know if this is necessary.

B3n30 replied:
Since Subv reports it works without the sleep I will skip it for now

JayFoxRox replied:
It might work for his platform.. - This is practically a busyloop.
We should fix this. At least we should have a reminder about it in an issue or something.

B3n30 commented on src/room_member.cpp:
```diff
@@ -0,0 +1,210 @@
+
+void RoomMember::ReceiveLoop() {
+    // Receive packets while the connection is open
+    while (IsConnected()) {
```
In the Raknet examples a Sleep(30 / u_sleep(30*1000) is used to "keep Raknet responsive". I don't know if this is necessary.

B3n30 replied:
Will skip this, too

B3n30 commented on include/tunnler/assert.h:
```diff
@@ -0,0 +1,51 @@
+// Copyright 2013 Dolphin Emulator Project / 2014 Citra Emulator Project
```
Should we add tunnler. At least its 2017... :D

JayFoxRox replied:
imo: Nope, it's not our file

B3n30 commented on include/tunnler/room.h:
```diff
@@ -0,0 +1,121 @@
+        RakNet::AddressOrGUID network_address; ///< The network address of the remote peer.
+    };
+
+    using MemberList = std::vector<Member>;
```
We now have Room::MemberList and RoomMember::MemberList, both holding almost the same information. I know technically there is no problem with that, but going to the code sometimes it's hard to distinguish which one is meant. Maybe rename one to ServerMemberList or similar or use a std::map<RakNet::AddressOrGUID, RoomMember::MemberInformation> since we have to store the complete MemberInformation on the Server anyway.

JayFoxRox replied:
I didn't read the code yet [just your review comments], but introducing "Server" would be even more confusing.
The general concept of naming these functions was that a non-network person or even layman would get these terms - this will also be reflected by the GUI.

B3n30 commented on include/tunnler/room_member.h:
```diff
@@ -108,31 +113,45 @@ class RoomMember final {
-    MemberInformation member_information;
-    RoomInformation room_information;
+    std::atomic<State> state; ///< Current state of the RoomMember.
+    MemberList member_information; ///< Information about the clients connected to the same room as us.
```
member_information_list ?

JayFoxRox replied:
I think the current name is fine: "information" is non-countable. while this doesn't imply a plural from the naming, it does from the context (and so does the comment [which should probably be renamed to say ~"all room-members in the joined room" instead of client])

B3n30 commented on include/tunnler/tunnler.h:
```diff
@@ -6,10 +6,14 @@
 using MacAddress = std::array<uint8_t, 6>;
 
+/// A special MAC address that tells the room we're joining to assign us a MAC address automatically.
+const MacAddress NoPreferredMac = { 0, 0, 0, 0, 0, 0 };
```
Maybe just use the broadcast mac for that, since this should be the only mac address a client shouldn't have. 0:0:0:0:0:0 is as far as i know a legit mac address.

B3n30 commented on include/tunnler/room_member.h:
```diff
@@ -108,31 +113,45 @@ class RoomMember final {
+    RakNet::RakPeerInterface* peer; ///< RakNet network interface.
 
-    RakNet::RakPeerInterface* peer;   // RakNet network interface
+    std::string nickname; ///< The nickname of this member.
```
Maybe just store MemberInformation for this client, thus MAC adress is stored there too.

JayFoxRox replied:
Not sure. I feel that the MemberInformation / MemberList stuff is a user-facing (API) based structure.

Consider network activity: If we count network activity (sent WifiPackets / s per RoomMember) to show in the frontend* we'd suddenly have useless fields stored in the RoomMember.

Also it adds a layer of indirection which might make the code slightly more complicated for no reason.

*(which I'm intending so network spammers could be found more easily)

B3n30 commented on src/room.cpp:
```diff
@@ -0,0 +1,188 @@
+// Refer to the license.txt file included.
+
+#include "tunnler/room.h"
+
```
unnecessary space

JayFoxRox replied:
lol LC?!

B3n30 replied:
sorry. I meant unnecessary line :-P

JayFoxRox replied:
I get that - but this is something lioncash (= LC) would say during review (and LC reviews *everything*).
He's very "uptight" (for a lack of a better term) about keeping code short by removing useless empty lines :P

B3n30 commented on src/room.cpp:
```diff
@@ -0,0 +1,188 @@
+}
+
+bool Room::IsValidMacAddress(const MacAddress& address) {
+    // A MAC address is valid if it is not already taken by anybody else in the room.
```
If my suggestion about NoPreferredMac=FF:FF:FF:FF:FF:FF isn't considered here should't be a check that address not equals the broadcast address. I can't think of any other special macs but broadcast shoudn't be vaild

JayFoxRox replied:
resolved

B3n30 commented on src/room_member.cpp:
```diff
@@ -0,0 +1,210 @@
+}
+
+RoomMember::~RoomMember() {
+    ASSERT_MSG(!IsConnected(), "RoomMember is being destroyed while connected");
```
Why not just call Leave? Since this destructor my be called even if I'm connected (for example just closing citra while connected)

JayFoxRox replied:
imo it should be part of the Application / Citra cleanup to worry about these things.
The current behaviour is fine

JayFoxRox replied:
resolved

B3n30 commented on include/tunnler/room.h:
```diff
@@ -0,0 +1,121 @@
+    using MemberList = std::vector<Member>;
+
+    Room() { }
+    ~Room() { }
```
At least Destroy() should be called here

JayFoxRox replied:
same as for RoomMember: Leave this kind of behaviour to applications

B3n30 replied:
Imo this couldn't be handled by the applications, since the destructor is called "automatically" in some cases. Thus closing the room/room_member and sending the disconnect message should be handled...

B3n30 reviewed (changes requested):
I might not got everything.
Some are just questions or notes to think about

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -0,0 +1,210 @@
+    }
+}
+
+std::deque<WifiPacket> RoomMember::PopWifiPackets(WifiPacket::PacketType type, const MacAddress& mac_address) {
```
Given B3N30s latest comment: It should only be sensical that filtering for `mac_address == NoPreferredMac / BroadcastMac` should disable the filter action (for mac only). Keep this in mind for the logic and documentation.

(Add a TODO?)

Subv replied:
The to-do is already there, see L73

JayFoxRox commented:
We are discussing minuscule things by now. I feel like this has been discussed enough. 
So I suggest keeping "nits" to yourself at this point and only address actual design issues.

@Subv please handle the latest review comments; then make sure this code is compilable so we can merge this to master asap (just handle the remaining review comments). We should have a stable master as base to work on for more important things.

(If remaining design or code issues are found we can discuss them in GitHub issues or add comments after the merge)

JayFoxRox commented on CMakeLists.txt:
```diff
@@ -1,6 +1,10 @@
 
+if(UNIX)
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -std=gnu++14")
+endif()
```
Any reason we don't use the cmake way of doing this?
(or use whatever Citra uses :P )

B3n30 replied:
Only my stupidity and my inability of writing cmake. I will just look what citra uses and add it here

B3n30 replied:
addressed

JayFoxRox commented on src/room.cpp:
```diff
@@ -37,8 +37,6 @@ void Room::Destroy() {
         server = nullptr;
     }
-
-    room_thread->join();
```
Not sure about this change, I guess it's fine, but I feel like we should continue handling packets even during shutdown.
Thread closure (setting state + joining thread) should probably happen after shutdown, before destroying the instance.

//edit: I'll add an issue once we merged this [or do it now, it fits with the next PR]

B3n30 replied:
I think since we close the room, receiving packets isn't needed any more, and the way I arranged it, allows the room/server to work without a mutex

JayFoxRox replied:
In case of RoomMember we'd still get confirmation that we successfully disconnected etc. (I believe)
I guess the same will hold true for Room.

Subv replied:
This order of things is actually ok, the thread will keep receiving packets until there's nothing more to receive, then exit the inner loop and then exit the outer loop (because state = Closed), after that we shut down the instance (which should tell the clients that we're shutting down, if i'm reading the docs correctly)

JayFoxRox commented:
This looks good now, basicly this C14 thing and an edge-case. Merge please :)
We can then do follow up PRs

B3n30 commented:
For me LGTM...

B3n30 committed:
removed Room::server_lock by rearanging in Room::Destroy()

Subv commented:
LGTM, let's merge it and iterate on it with further PRs
